content/post/flinksql-time.adoc

```diff
@@ -0,0 +1,1033 @@
+At 9am a supermarket opens and I buy a tin of baked beans.
+At 6pm the store closes, the end-of-day runs on the in-store PoS, and sends the data back to head office.
+Perhaps at 10pm all the data from all the stores has been received and processing starts.
+
+In the modern age many platforms will be much lower latency.
+PoS systems would be sending information about my tin of baked beans purchase back to a central system (probably using Apache Kafka…) as soon as it happened.
+But even then, times can differ.
+There's the natural latency that the speed of light has a strong opinion on, on top of which is any other processing or batching that might be introduced along the way.
+Then there are power outages and network blips and gremlins that can mean records can be significantly delayed.
+
+There's a whole category of data that wasn't a significant thing in the past—data that's generated from mobile devices.
+Even with improving cellular coverage and in-flight wifi, data can still be delayed by minutes or hours.
+
+So that's event time and processing time. But there are other types of time too:
+
+* The time the record was written in the system
+* The time a record was first created
+* The time a record was last updated
+* Arbitrary time fields on a record, such as an order record with times that the order was place, fulfilled, and shipped
+
+I'll come back to these in detail later.
+For now, just keep in mind that there is no single "correct" time to use.
+It depends on what your processing is supposed to be doing.
+If you are trying to calculate how many orders were shipped in the last hour, then using the time that the order was created or that it landed in the processing system would give you _an_ answer—it would just be the wrong one.
+
+TIP: https://www.oreilly.com/radar/the-world-beyond-batch-streaming-101/[This seminal article from Tyler Akidau] is nearly ten years old, but is still a highly relevant and excellent read and covers a lot of what you need to know about time in stream processing.
+
+== Time in Kafka
+
+Each Kafka message is made up of https://kafka.apache.org/documentation/#record[several parts]:
+
+. Key
+. Value (the payload)
+. Header
+. Timestamp
+
+Here's an example record.
+The value is in JSON, which I've pretty-printed to make is easier to follow:
+
+[source,javascript]
+----
+Key (14 bytes): {"order_id":1}
+Value (205 bytes):
+    {
+        "order_id": 1,
+        "customer_id": 1001,
+        "total_amount": 149.99,
+        "status": "pending",
+        "created_at": "2025-04-25 09:44:25",
+        "shipped_at": null,
+        "shipping_address": "221B Baker Street, London",
+        "payment_method": "Credit Card"
+    }
+Timestamp: 1745587589625
+----
+
+There are two times here:
+
+. The _event time_ which we'll take as `created_at` from the value: `2025-04-25 09:44:25`.
+. The _processing time_ which is stored in the Kafka message timestamp `1745587589625`.
```

Comment:
I'm uncomfortable calling this _processing time_, since this is a different meaning of processing time than the one used by Flink. A Flink time attribute based on processing time will have a different value each time it is accessed -- and cannot be based on a stored value.

Reply:
Yeah, this scratched me too. It's more like an "ingestion time" in Kafka?

Taking a step back, I wonder whether that entire discussion of Kafka is not too much of a distraction?